Performance Enhancement :Enable Passing of requestPath to CRT and monitor the progress using CRTs S3MetaRequestProgress

joviegas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
- The File IO operation during Upload of the File should not be done in SDK side.
- The requestPath is passed to the CRT and CRT does the IO operation.
 - CRT side changes  https://github.com/awslabs/aws-crt-java/pull/595 and https://github.com/awslabs/aws-c-s3/pull/344 


## Modifications
<!--- Describe your changes in detail -->
- Removed the code where S3CrtRequestBodyStreamAdapter is passed to CRTHttp client
- Now we just pass null is request Path is passed to the S3MetaRequest.
- The Progress listener will listen to S3MetaRequestProgress from CRT



## Testing
- No need of new Test since this is internal behaviour test, the existing test cases should cover this.
- Executed Integ tests.
<!--- Please describe in detail how you tested your changes -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## License
<!--- The SDK is released under the Apache 2.0 license (http://aws.amazon.com/apache2.0/), so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a Contributor License Agreement (http://en.wikipedia.org/wiki/Contributor_License_Agreement) -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [x] I confirm that this pull request can be released under the Apache 2 license

millems commented on services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/CrtS3TransferManager.java:
```diff
@@ -65,29 +65,25 @@ public FileUpload uploadFile(UploadFileRequest uploadFileRequest) {
 
             CompletableFuture<PutObjectResponse> crtFuture =
-                s3AsyncClient.putObject(putObjectRequest, requestBody);
+                s3AsyncClient.putObject(putObjectRequest, AsyncRequestBody.fromFile(uploadFileRequest.source()));
```
If CRT is reading directly from disk, does this publisher get used?

millems commented on services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtProgressListener.java:
```diff
@@ -0,0 +1,80 @@
+import software.amazon.awssdk.core.async.listener.PublisherListener;
+import software.amazon.awssdk.crt.s3.S3MetaRequestProgress;
+
+/**
+ * S3CrtProgressListener delegates events to the underlying delegateListener if defined, avoiding null checks for API when using
+ * S3MetaRequestProgress for GET calls.
+ */
```
This feels like a weird application of polymorphism. Instead of S3CrtProgressListener, can we just have a NoOpPublisherListener that we create if delegateListener is null? That just means we need the null check in the setup code of S3CrtResponseHandlerAdapter instead of per-method here.

millems commented on services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3InternalSdkHttpExecutionAttribute.java:
```diff
@@ -41,6 +43,13 @@ public final class S3InternalSdkHttpExecutionAttribute<T> extends SdkHttpExecuti
     public static final S3InternalSdkHttpExecutionAttribute<Region> SIGNING_REGION =
         new S3InternalSdkHttpExecutionAttribute<>(Region.class);
 
+    public static final S3InternalSdkHttpExecutionAttribute<Path> SOURCE_REQ_PATH =
+        new S3InternalSdkHttpExecutionAttribute<>(Path.class);
+
+
+    public static final S3InternalSdkHttpExecutionAttribute<PublisherListener> CRT_PROGRESS_LISTENER =
+        new S3InternalSdkHttpExecutionAttribute<>(PublisherListener.class);
```
This is marked as @SdkInternalApi, so it can't be used by the transfer manager module, like we're doing above. Do we need to move a subset to a @SdkProtectedApi?

joviegas replied:
Done

millems commented on services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3InternalSdkHttpExecutionAttribute.java:
```diff
@@ -41,6 +43,13 @@ public final class S3InternalSdkHttpExecutionAttribute<T> extends SdkHttpExecuti
     public static final S3InternalSdkHttpExecutionAttribute<Region> SIGNING_REGION =
         new S3InternalSdkHttpExecutionAttribute<>(Region.class);
 
+    public static final S3InternalSdkHttpExecutionAttribute<Path> SOURCE_REQ_PATH =
+        new S3InternalSdkHttpExecutionAttribute<>(Path.class);
```
Naming nitpicks, since this actually seems like a protected API.
* `SOURCE` - Do we ever expect to use this for something other than a PutObject - like a GetObject, where it's actually the `DESTINATION`?
* `REQ` - We should almost always use the full word for something instead of shortening it, so `REQUEST`. It looks like a lot of the existing properties are request properties, though, so I don't think REQUEST is necessary.

If it's always the source, what about `PUT_OBJECT_PATH`? `UPLOAD_PATH`?
If it's potentially also destination, what about `UPLOAD_DOWNLOAD_PATH`?

joviegas replied:
Thanks Matt..
Named it as OBJECT_FILE_PATH so that it can be used for any file path for get/put

millems reviewed (commented):
* Is it possible to test that what we're intending to fix actually got fixed?
* Is there performance testing to verify that this change didn't hurt performance?
* Do we already have tests that assert the progress listener changes were backwards compatible?

joviegas commented:
> * Is it possible to test that what we're intending to fix actually got fixed?
> 
Thanks Matt for this suggestion , tested this and realized the CRT doesnot actually fix the issue where upload file size is more than RAM size. I have cut a different ticket to CRT. However , this fix is still needed for performance enhancement. Removed the mention of the orginal issue.

> * Is there performance testing to verify that this change didn't hurt performance?
> 
Yes , I saw that the latencies was reduced to half , I have updated the internal doc.

> * Do we already have tests that assert the progress listener changes were backwards compatible?
> 
Yes , we already have test .

> * Are there any path-processing nuances that may have changed when we switched to the CRT? E.g. handling of the current working directory or expansion of ~/?

No.

zoewangg commented on services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java:
```diff
@@ -46,9 +48,14 @@ public final class S3CrtResponseHandlerAdapter implements S3MetaRequestResponseH
     private volatile S3MetaRequest metaRequest;
 
-    public S3CrtResponseHandlerAdapter(CompletableFuture<Void> executeFuture, SdkAsyncHttpResponseHandler responseHandler) {
+    private final PublisherListener<S3MetaRequestProgress> progressListener;
```
Do we need to invoke `this.progressListener.subscriberOnError();` in `onErrorResponseComplete`?

zoewangg replied:
`failResponseHandlerAndFuture` doesn't seem to be invoked from `onErrorResponseComplete` if it successfully marshalls the error response though. Can we add a test to verify it? [This test](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/S3TransferManagerListenerTest.java#L65) may not suffice now that we are relying on the real CRT client to report progress and it is using a mock s3 client.

```
        responsePublisher.send(ByteBuffer.wrap(errorPayload))
                         .thenRun(responsePublisher::complete)
                         .handle((ignore, throwable) -> {
                             if (throwable != null) {
                                 failResponseHandlerAndFuture(throwable);
                                 return null;
                             }
                             completeFutureAndCloseRequest();
                             return null;
                         });
```

joviegas replied:
Added new mock tests and removed  explicit onError

zoewangg commented on services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/CrtContentLengthOnlyAsyncFileRequestBody.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Override
+    public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
+        subscriber.onSubscribe(new Subscription() {
```
It doesn't seem to comply with the spec
>Subscription.request and Subscription.cancel MUST only be called inside of its Subscriber context.

We can probably use [NoopSubscription](https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/util/NoopSubscription.java) here. We need to move it outside of the internal package though since it's used cross module

```

            subscriber.onSubscribe(new NoopSubscription(s));
            subscriber.onError(new UnsupportedOperationException());
```

Alternatively, we can follow the same logic, i.e., calling onError in `request` and no-op in cancel.

joviegas replied:
Will use `Alternatively` since its better to fail fast in such cases since its error if we try to subscribe CrtContentLengthOnlyAsyncFileRequestBody

zoewangg replied:
Oops, apologies, I mistyped, I meant we can follow the same logic in this class if we don't want to make NoopSubscription protected. We have to call onSubscribe first per reactive streams spec.

zoewangg commented on services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/S3CrtClientPutObjectIntegrationTest.java:
```diff
@@ -86,6 +90,25 @@ void putObject_fileRequestBody_objectSentCorrectly() throws Exception {
+
+
+    @Test
+    void putObject_ContentOnlyAsynRequestBody_fails()  {
```
Nvm, got it

joviegas replied:
making sure that when CrtContentLengthOnlyAsyncFileRequestBody is used and if we try to subscribe it then it will fail

joviegas commented on services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3TmInternalSdkHttpExecutionAttribute.java:
```diff
@@ -0,0 +1,37 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.internal.crt;
```
done

joviegas commented on services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3TmInternalSdkHttpExecutionAttribute.java:
```diff
@@ -0,0 +1,37 @@
+import software.amazon.awssdk.http.SdkHttpExecutionAttribute;
+
+@SdkProtectedApi
+public class S3TmInternalSdkHttpExecutionAttribute<T> extends SdkHttpExecutionAttribute<T> {
```
done

zoewangg commented on services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/progress/TransferProgressUpdater.java:
```diff
@@ -107,8 +107,7 @@ public void publisherSubscribe(Subscriber<? super S3MetaRequestProgress> subscri
             public void subscriberOnNext(S3MetaRequestProgress s3MetaRequestProgress) {
-                incrementBytesTransferred(Math.toIntExact(s3MetaRequestProgress.getBytesTransferred()),
-                                          s3MetaRequestProgress.getContentLength());
+                incrementBytesTransferred(Math.toIntExact(s3MetaRequestProgress.getBytesTransferred()));
```
We probably should not cast it to int. Can we change incrementBytesTransferred to take a long instead?
